packages/core/src/view/provider.ts

```diff
@@ -256,8 +256,24 @@ function _createProviderInstance(view: ViewData, def: NodeDef): any {
       return createClass(
           view, def.parent!, allowPrivateServices, providerDef!.value, providerDef!.deps);
     case NodeFlags.TypeFactoryProvider:
-      return callFactory(
+      let injectable: any;
+      injectable = callFactory(
           view, def.parent!, allowPrivateServices, providerDef!.value, providerDef!.deps);
+      // Update def flags if factory injectable has ngOnDestroy function.
+      if (injectable != null && typeof injectable === 'object' &&
+          !(def.flags & NodeFlags.OnDestroy) && typeof injectable.ngOnDestroy === 'function') {
+        def.flags |= NodeFlags.OnDestroy;
+        view.def.nodeFlags |= NodeFlags.OnDestroy;
+        if (def.parent) {
+          let parent: NodeDef|null = def.parent;
+          while (parent !== null) {
```

Comment:
I just realized we may exit the loop early if possible:

```suggestion
          while (parent !== null && !(parent.childFlags & NodeFlags)) {
```

Reply:
well seen!